### ISSUE TITLE
Failing to use Compose runtime in common code targeting native

### ISSUE THREAD

**Issue by kirillzh** [JetBrains/compose-multiplatform#2346]
Failing to compile simple Compose function targeting native code (iOS `iosX64` or `iosArm64`).

Kotlin `1.7.10`
Jetbrains Compose `1.2.0-beta01`
Compose compiler `1.3.0`

Repo code is in this branch: https://github.com/kirillzh/kmm-production-sample/tree/common-compose.

```kotlin
// shared/src/commonMain
@Composable
fun Foo() {
}
```

Error:
```
> Task :shared:linkDebugFrameworkIosSimulatorArm64 FAILED
w: Mimalloc allocator isn't supported on target ios_simulator_arm64. Used standard mode.
e: Compilation failed: No file for com.github.jetbrains.rssreader/Foo|4071909078632269291[0] (@ com.github.jetbrains.rssreader/Foo|4071909078632269291[0]) in module `<RssReader:shared>`[ModuleDescriptorImpl@57d43629]

 * Source files: 
 * Compiler version info: Konan: 1.7.10 / Kotlin: 1.7.20
 * Output kind: FRAMEWORK

e: java.lang.IllegalStateException: No file for com.github.jetbrains.rssreader/Foo|4071909078632269291[0] (@ com.github.jetbrains.rssreader/Foo|4071909078632269291[0]) in module `<RssReader:shared>`[ModuleDescriptorImpl@57d43629]
	at org.jetbrains.kotlin.backend.common.serialization.BasicIrModuleDeserializer.deserializeIrSymbol(BasicIrModuleDeserializer.kt:95)
```

<details>

  <summary>Full Error</summary>

```
> Task :shared:linkDebugFrameworkIosSimulatorArm64 FAILED
w: Mimalloc allocator isn't supported on target ios_simulator_arm64. Used standard mode.
e: Compilation failed: No file for com.github.jetbrains.rssreader/Foo|4071909078632269291[0] (@ com.github.jetbrains.rssreader/Foo|4071909078632269291[0]) in module `<RssReader:shared>`[ModuleDescriptorImpl@57d43629]

 * Source files: 
 * Compiler version info: Konan: 1.7.10 / Kotlin: 1.7.20
 * Output kind: FRAMEWORK

e: java.lang.IllegalStateException: No file for com.github.jetbrains.rssreader/Foo|4071909078632269291[0] (@ com.github.jetbrains.rssreader/Foo|4071909078632269291[0]) in module `<RssReader:shared>`[ModuleDescriptorImpl@57d43629]
	at org.jetbrains.kotlin.backend.common.serialization.BasicIrModuleDeserializer.deserializeIrSymbol(BasicIrModuleDeserializer.kt:95)
	at org.jetbrains.kotlin.backend.common.serialization.IrModuleDeserializer.declareIrSymbol(IrModuleDeserializer.kt:74)
	at org.jetbrains.kotlin.backend.common.serialization.KotlinIrLinker.findDeserializedDeclarationForSymbol(KotlinIrLinker.kt:127)
	at org.jetbrains.kotlin.backend.common.serialization.KotlinIrLinker.deserializeOrResolveDeclaration(KotlinIrLinker.kt:169)
	at org.jetbrains.kotlin.backend.common.serialization.KotlinIrLinker.getDeclaration(KotlinIrLinker.kt:158)
	at org.jetbrains.kotlin.ir.util.ExternalDependenciesGeneratorKt.getDeclaration(ExternalDependenciesGenerator.kt:57)
	at org.jetbrains.kotlin.ir.util.ExternalDependenciesGenerator.generateUnboundSymbolsAsDependencies(ExternalDependenciesGenerator.kt:44)
	at org.jetbrains.kotlin.psi2ir.generators.ModuleGenerator.generateUnboundSymbolsAsDependencies(ModuleGenerator.kt:52)
	at org.jetbrains.kotlin.psi2ir.Psi2IrTranslator.generateModuleFragment(Psi2IrTranslator.kt:92)
	at org.jetbrains.kotlin.psi2ir.Psi2IrTranslator.generateModuleFragment$default(Psi2IrTranslator.kt:75)
	at org.jetbrains.kotlin.backend.konan.PsiToIrKt.psiToIr(PsiToIr.kt:187)
	at org.jetbrains.kotlin.backend.konan.ToplevelPhasesKt$psiToIrPhase$1.invoke(ToplevelPhases.kt:120)
	at org.jetbrains.kotlin.backend.konan.ToplevelPhasesKt$psiToIrPhase$1.invoke(ToplevelPhases.kt:118)
	at org.jetbrains.kotlin.backend.common.phaser.PhaseBuildersKt$namedOpUnitPhase$1.invoke(PhaseBuilders.kt:96)
	at org.jetbrains.kotlin.backend.common.phaser.PhaseBuildersKt$namedOpUnitPhase$1.invoke(PhaseBuilders.kt:94)
	at org.jetbrains.kotlin.backend.common.phaser.NamedCompilerPhase.invoke(CompilerPhase.kt:96)
	at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:29)
	at org.jetbrains.kotlin.backend.common.phaser.NamedCompilerPhase.invoke(CompilerPhase.kt:96)
	at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.invokeToplevel(CompilerPhase.kt:43)
	at org.jetbrains.kotlin.backend.konan.KonanDriverKt.runTopLevelPhases(KonanDriver.kt:40)
	at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:94)
```

</details>

Moving Composable function under a class/interface, gives a bit different error:

```kotlin
// shared/src/commonMain
interface Example {
  @Composable
  fun Foo() {
  }
}
```

Error:
```
e: java.lang.AssertionError: Unbound symbols not allowed

Unbound public symbol IrSimpleFunctionPublicSymbolImpl: com.github.jetbrains.rssreader/Example.Foo|4071909078632269291[0]
	at org.jetbrains.kotlin.ir.util.SymbolTableKt.noUnboundLeft(SymbolTable.kt:1141)
	at org.jetbrains.kotlin.psi2ir.Psi2IrTranslator.generateModuleFragment(Psi2IrTranslator.kt:96)
	at org.jetbrains.kotlin.psi2ir.Psi2IrTranslator.generateModuleFragment$default(Psi2IrTranslator.kt:75)
	at org.jetbrains.kotlin.backend.konan.PsiToIrKt.psiToIr(PsiToIr.kt:187)
	at org.jetbrains.kotlin.backend.konan.ToplevelPhasesKt$psiToIrPhase$1.invoke(ToplevelPhases.kt:120)
	at org.jetbrains.kotlin.backend.konan.ToplevelPhasesKt$psiToIrPhase$1.invoke(ToplevelPhases.kt:118)
	at org.jetbrains.kotlin.backend.common.phaser.PhaseBuildersKt$namedOpUnitPhase$1.invoke(PhaseBuilders.kt:96)
	at org.jetbrains.kotlin.backend.common.phaser.PhaseBuildersKt$namedOpUnitPhase$1.invoke(PhaseBuilders.kt:94)
	at org.jetbrains.kotlin.backend.common.phaser.NamedCompilerPhase.invoke(CompilerPhase.kt:96)
```

**Comment by DjakaTechnology** [comment]
Possible related to https://github.com/JetBrains/compose-jb/issues/2342

**Comment by gregriggins36** [comment]
I'm getting the exact same issue. Just adding `@Composable` is bringing this error

**Comment by dima-avdeev-jb** [comment]
Can you please try version **1.3.0-alpha01-dev849**
with kotlin.native.cacheKind=none

**Comment by gregriggins36** [comment]
> I'm getting the exact same issue. Just adding `@Composable` is bringing this error

I fixed my issue by making the `@Composable` function `internal` 🤦‍♂️

**Comment by dima-avdeev-jb** [comment]
Maybe relates with https://github.com/JetBrains/compose-jb/issues/2386

**Comment by kevincianfarini** [comment]
I'm also getting this issue, which was also fixed by marking my composable internal!

**Comment by kirillzh** [comment]
@dima-avdeev-jb, I'm understanding that it's not currently possible to make a Composable function public in common code while targeting native. 

This is blocking us from being able to modularize the code with Composables, and most importantly being able to incrementally migrate Android logic into shared code base incrementally.

Is there a fix for this in some dev release, and if not, would you have an estimate for when this is planned to be fixed?

**Comment by dima-avdeev-jb** [comment]
@kirillzh Yes, you can't use public Composable functions in iOS for now.
But, you can use **internal** functions.

**Comment by TaehoonLeee** [comment]
@kirillzh Except for the composable function of the module finally used in iOS, it is okay to use the public composable function.

For example, I use feature-wise modular like below
```mermaid
graph TD;
A[feature1] --> C[presentation]
B[feature2] --> C
```
In this case, public composable functions in feature1 and feature2 are available.

**Comment by pablichjenkov** [comment]
Hi @TaehoonLeee, could you share a simple example of what you described. I tried on my side and still no way to expose a Composable in iOS target.
This feature is important since I want to have a shared-ui library, made of components that other projects consume as Composables.

**Comment by TaehoonLeee** [comment]
@pablichjenkov 

It is impossible to use composable functions directly on iOS target. Therefore, all the composable functions in the module exported to the iOS target must be specified as internal.

However, the composable functions in other modules may be public. Using this point, it seems possible to modularize composable function.

https://github.com/TaehoonLeee/multi-module-clean-architecture/tree/multi-platform

In above project the iOS app uses presentation module only. 
So the composable functions written in the presentation module should be internal. And the others in the other modules are can be public.

My English is limited. I ask for your understanding.

**Comment by pablichjenkov** [comment]
@TaehoonLeee ,
Don't worry about the English mine is bad too.
I checked your project and it is similar to jetbrains experimental examples. I understand what you are talking about. And yes, I am aware I can expose a UIViewController from the share module, Presentation in your App. But that is not my use case. 
My use case is that the Presentation module should expose Composables, to be embedded in other modules Composables. If Presentation exposes a non-composable I cannot embed it inside  another iosMain.kt in a consumer project.
Think about it as a ui-library that will be consumed by another ui-library.

**Comment by kevincianfarini** [comment]
Do we know if this is resolved by compose compiler 1.4.0 and Kotlin 1.8.20? I'm seeing similar mentions in other related issues such as https://github.com/JetBrains/compose-multiplatform/issues/2901#issuecomment-1504894914

**Comment by pablichjenkov** [comment]
> Do we know if this is resolved by compose compiler 1.4.0 and Kotlin 1.8.20? I'm seeing similar mentions in other related issues such as https://github.com/JetBrains/compose-multiplatform/issues/2901#issuecomment-1504894914

Yeup, got fixed in 1.4.0

**Comment by kevincianfarini** [comment]
Wohoo!

**Comment by dima-avdeev-jb** [comment]
Fixed in Compose 1.4.0

**Comment by okushnikov** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/CMP-2346">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.
